### PR TITLE
camera enhancements added

### PR DESCRIPTION
**Pull request by VikashHart**
[Link to Ticket 🎟](_paste_link_here_)

## Description
camera enhancements added:
1. pinch to zoom
2. flash
3. tap to focus
4. double tap to switch camera

## Type of change

[//]: # (Please delete options that are not relevant.)

- [ ] 🐞Bug fix (non-breaking change which fixes an issue)
- [x] 🌟New feature (non-breaking change which adds functionality)
- [ ] 🚧Breaking change (fix or feature that would cause existing functionality to not work as expected)

## UI Changes 
[//]: # (Add images/gifs if applicable)

Before            | After
----------------  | -----------------
image/gif before  |  image / gif after

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Manually tested
- [ ] Unit tests added/updated

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Update my ticket with relevant information (inProgress or any comments needed)

### CODE REVIEW

File: PhotoBooth/Camera Page/CameraViewController.swift [hunk 1]
```diff
@@ -172,9 +179,51 @@ class CameraViewController: UIViewController {
...
+        switch viewModel.flashType {
+        case .on:
+            UIView.animate(withDuration: 0.1,
+                           animations: {
+                            self.flashButton.transform = CGAffineTransform(scaleX: 0.6, y: 0.6)
+                            self.viewModel.toggleFlash()
+            },
+                           completion: { _ in
+                            UIView.animate(withDuration: 0.1) {
+                                print(self.viewModel.captureSession.getFlashMode().rawValue)
```
Comment (vhart) [inline PhotoBooth/Camera Page/CameraViewController.swift]: Left in a print statement here

File: PhotoBooth/Camera Page/CameraViewController.swift [hunk 2]
```diff
@@ -172,9 +179,51 @@ class CameraViewController: UIViewController {
...
+                                self.flashButton.setImage(UIImage(named: "flash_on"), for: .normal)
+                                self.flashButton.transform = CGAffineTransform.identity
+                            }
+            })
+        default:
```
Comment (vhart) [inline PhotoBooth/Camera Page/CameraViewController.swift]: what about `.auto`?

File: PhotoBooth/Camera Page/CameraViewController.swift [hunk 3]
```diff
@@ -172,9 +179,51 @@ class CameraViewController: UIViewController {
...
+        case .off:
+            UIView.animate(withDuration: 0.1,
+                           animations: {
+                            self.flashButton.transform = CGAffineTransform(scaleX: 0.6, y: 0.6)
+                            self.viewModel.toggleFlash()
+            },
+                           completion: { _ in
+                            UIView.animate(withDuration: 0.1) {
+                                print(self.viewModel.captureSession.getFlashMode().rawValue)
+                                self.flashButton.setImage(UIImage(named: "flash_on"), for: .normal)
```
Comment (vhart) [inline PhotoBooth/Camera Page/CameraViewController.swift]: these two animation blocks are the same. the only difference is the image. Can you use the the switch statement to choose an image and then do the animation in one go?

```swift
var flashImage: UIImage?
switch {
  set flashImage
}

UIView.animate(.... {
 completion: {

       self.flashButton.setImage(flashImage)
       ....
   }
})
```

File: PhotoBooth/Camera Page/CameraViewController.swift [hunk 4]
```diff
@@ -187,11 +236,63 @@ class CameraViewController: UIViewController {
             presentCancelationPartialModal()
         }
     }
+
+    @objc func handleSingleTap(tapGesture: UITapGestureRecognizer) {
+        let point = tapGesture.location(in: self.previewLayerContainer)
+        print("current point: \(point)")
```
Comment (vhart) [inline PhotoBooth/Camera Page/CameraViewController.swift]: print statement

File: PhotoBooth/Camera Page/CameraViewController.swift [hunk 5]
```diff
@@ -187,11 +236,63 @@ class CameraViewController: UIViewController {
...
+
+    @objc func pinch(_ pinch: UIPinchGestureRecognizer) {
+        let newScaleFactor = viewModel.captureSession.minMaxZoom(pinch.scale * viewModel.zoomFactor)
+
+        switch pinch.state {
+        case .began: fallthrough
+        case .changed: viewModel.captureSession.updateZoomScaleFactor(scale: newScaleFactor)
+        case .ended:
+            viewModel.updateZoomFactor(factor: newScaleFactor)
+            viewModel.captureSession.updateZoomScaleFactor(scale: viewModel.zoomFactor)
```
Comment (vhart) [inline PhotoBooth/Camera Page/CameraViewController.swift]: if the capture session updates it zoom factor does the view model also need to track it?

File: PhotoBooth/Camera Page/CameraViewController.swift (hunk 2 above)
Comment (VikashHart) [reply PhotoBooth/Camera Page/CameraViewController.swift]: you said we weren't going with auto. you said to make it like snapchats

File: PhotoBooth/Camera Page/CameraViewController.swift (hunk 5 above)
Comment (VikashHart) [reply PhotoBooth/Camera Page/CameraViewController.swift]: yea the factor here is for the views zoom factor so we can see the visible change and the capture session method sets it so the photo will be captured with the right zoom scale

File: PhotoBooth/Camera Page/CameraViewController.swift (hunk 2 above)
Comment (vhart) [reply PhotoBooth/Camera Page/CameraViewController.swift]: but you still have auto as an option

File: PhotoBooth/Camera Page/CameraViewController.swift (hunk 5 above)
Comment (vhart) [reply PhotoBooth/Camera Page/CameraViewController.swift]: Right but do we need both? In other words why does the view model need to know the zoom scale if the capture session already has it?

File: PhotoBooth/Camera Page/CameraViewController.swift (hunk 2 above)
Comment (VikashHart) [reply PhotoBooth/Camera Page/CameraViewController.swift]: you mean in the set flash options?

File: PhotoBooth/Camera Page/CameraViewController.swift (hunk 3 above)
Comment (VikashHart) [reply PhotoBooth/Camera Page/CameraViewController.swift]: done

File: PhotoBooth/Camera Page/CameraViewController.swift (hunk 5 above)
Comment (VikashHart) [reply PhotoBooth/Camera Page/CameraViewController.swift]: I'm running into a problem. I can remove all of that but I can't get the zoom scale to go back to 1 after the shoot. the pinch was the thing modifying the scale but idk another way to set it back

File: PhotoBooth/Camera Page/CameraViewController.swift (hunk 2 above)
Comment (vhart) [reply PhotoBooth/Camera Page/CameraViewController.swift]: yeah

File: PhotoBooth/Components/Capture Session/CaptureSession.swift [hunk 6]
```diff
@@ -10,9 +10,30 @@ protocol PhotoCaptureable {
...
+    func minMaxZoom(_ factor: CGFloat) -> CGFloat
+
+    func updateZoomScaleFactor(scale factor: CGFloat)
+
     func updateOrientation(orientation: AVCaptureVideoOrientation)
 }
 
+enum FlashModeOption {
+    case on
+    case auto
```
Comment (vhart) [inline PhotoBooth/Components/Capture Session/CaptureSession.swift]: do we need auto here anymore?

File: PhotoBooth/Components/Capture Session/CaptureSession.swift [hunk 7]
```diff
@@ -10,9 +10,30 @@ protocol PhotoCaptureable {
...
+    func focus(touchLocation: CGPoint)
+
+    func minMaxZoom(_ factor: CGFloat) -> CGFloat
+
+    func updateZoomScaleFactor(scale factor: CGFloat)
+
     func updateOrientation(orientation: AVCaptureVideoOrientation)
 }
 
+enum FlashModeOption {
```
Comment (vhart) [inline PhotoBooth/Components/Capture Session/CaptureSession.swift]: what about `FlashStatus`?

File: PhotoBooth/Components/Capture Session/CaptureSession.swift [hunk 8]
```diff
@@ -10,9 +10,30 @@ protocol PhotoCaptureable {
...
+    case on
+    case auto
+    case off
+}
+
+struct CameraZoom {
```
Comment (vhart) [inline PhotoBooth/Components/Capture Session/CaptureSession.swift]: can this go in its own file?

File: PhotoBooth/Components/Capture Session/CaptureSession.swift [hunk 9]
```diff
@@ -53,12 +74,70 @@ class CaptureSession: NSObject, PhotoCaptureable, AVCapturePhotoCaptureDelegate
...
+            } catch {
+                // Handle errors here
+                print("\(error.localizedDescription)")
+            }
+        }
+    }
+
+    // Return zoom value between the minimum and maximum zoom values
+    func minMaxZoom(_ factor: CGFloat) -> CGFloat {
+        return min(min(max(factor, CameraZoom.min), CameraZoom.max), currentCamera!.activeFormat.videoMaxZoomFactor)
```
Comment (vhart) [inline PhotoBooth/Components/Capture Session/CaptureSession.swift]: Should we `!` here? I thinks its generally better to unwrap than bang.
`guard let camera = currentCamera else { return CameraZoom.min }`

File: PhotoBooth/Components/Capture Session/CaptureSession.swift [hunk 10]
```diff
@@ -53,12 +74,70 @@ class CaptureSession: NSObject, PhotoCaptureable, AVCapturePhotoCaptureDelegate
...
+    }
+
+    func updateZoomScaleFactor(scale factor: CGFloat) {
+        if let device = currentCamera {
+            do {
+                try device.lockForConfiguration()
+                defer { device.unlockForConfiguration() }
```
Comment (vhart) [inline PhotoBooth/Components/Capture Session/CaptureSession.swift]: you understand what defer is doing?

File: PhotoBooth/Camera Page/CameraViewController.swift (hunk 5 above)
Comment (vhart) [reply PhotoBooth/Camera Page/CameraViewController.swift]: Do we need it to go back after the shoot? Also Im not sure I understand what you mean by the pinch was the thing modifying the scale. Like I know that pinch modifies the scale but what does that have to do with after the shoot? I thought that was part of the reset method in the view model

File: PhotoBooth/Components/Capture Session/CaptureSession.swift (hunk 9 above)
Comment (VikashHart) [reply PhotoBooth/Components/Capture Session/CaptureSession.swift]: what do we return if there is no camera?

File: PhotoBooth/Components/Capture Session/CaptureSession.swift (hunk 8 above)
Comment (VikashHart) [reply PhotoBooth/Components/Capture Session/CaptureSession.swift]: what am I looking at?

File: PhotoBooth/Components/Capture Session/CaptureSession.swift (hunk 7 above)
Comment (VikashHart) [reply PhotoBooth/Components/Capture Session/CaptureSession.swift]: done

File: PhotoBooth/Components/Capture Session/CaptureSession.swift (hunk 6 above)
Comment (VikashHart) [reply PhotoBooth/Components/Capture Session/CaptureSession.swift]: done

File: PhotoBooth/Camera Page/CameraViewController.swift (hunk 5 above)
Comment (VikashHart) [reply PhotoBooth/Camera Page/CameraViewController.swift]: no the reset can't modify the visible scale

File: PhotoBooth/Components/Capture Session/CaptureSession.swift (hunk 10 above)
Comment (VikashHart) [reply PhotoBooth/Components/Capture Session/CaptureSession.swift]: as far as I understand the defer statement will always run last in the function

File: PhotoBooth/Components/Capture Session/CaptureSession.swift (hunk 8 above)
Comment (vhart) [reply PhotoBooth/Components/Capture Session/CaptureSession.swift]: `CameraZoom`

File: PhotoBooth/Components/Capture Session/CaptureSession.swift (hunk 9 above)
Comment (vhart) [reply PhotoBooth/Components/Capture Session/CaptureSession.swift]: thats what the else is for in the guard

File: PhotoBooth/Components/Capture Session/CaptureSession.swift (hunk 10 above)
Comment (vhart) [reply PhotoBooth/Components/Capture Session/CaptureSession.swift]: yep, just making sure it was intentional

File: PhotoBooth/Components/Capture Session/CaptureSession.swift (hunk 9 above)
Comment (VikashHart) [reply PhotoBooth/Components/Capture Session/CaptureSession.swift]: yes that's what I'm referring to

Comment (vhart) [reply PhotoBooth/Components/Capture Session/CaptureSession.swift]: wat?
the suggestion is to return CameraZoom.min if there is no camera

**Comment by VikashHart** [comment]
@vhart

**Pull request merged**